launch/launch/substitutions/boolean_substitution.py

```diff
@@ -114,144 +129,98 @@ def perform(self, context: LaunchContext) -> Text:
 
     def __init__(self, left: SomeSubstitutionsType, right: SomeSubstitutionsType) -> None:
-        """Create an OrSubstitution substitution."""
-        super().__init__()
-
-        self.__left = normalize_to_list_of_substitutions(left)
-        self.__right = normalize_to_list_of_substitutions(right)
-
-    @classmethod
-    def parse(cls, data: Sequence[SomeSubstitutionsType]
-              ) -> Tuple[Type['OrSubstitution'], Dict[str, Any]]:
-        """Parse `OrSubstitution` substitution."""
-        if len(data) != 2:
-            raise TypeError('and substitution expects 2 arguments')
-        return cls, {'left': data[0], 'right': data[1]}
-
-    @property
-    def left(self) -> List[Substitution]:
-        """Getter for left."""
-        return self.__left
-
-    @property
-    def right(self) -> List[Substitution]:
-        """Getter for right."""
-        return self.__right
+        """Create an AndSubstitution substitution."""
+        super().__init__(and_, left, right)
 
-    def describe(self) -> Text:
-        """Return a description of this substitution as a string."""
-        return f'AndSubstitution({self.left} {self.right})'
 
-    def perform(self, context: LaunchContext) -> Text:
-        """Perform the substitution."""
-        try:
-            left_condition = perform_typed_substitution(context, self.left, bool)
-        except (TypeError, ValueError) as e:
-            raise SubstitutionFailure(e)
-        try:
-            right_condition = perform_typed_substitution(context, self.right, bool)
-        except (TypeError, ValueError) as e:
-            raise SubstitutionFailure(e)
-
-        return str(left_condition or right_condition).lower()
+@expose_substitution('or')
+class OrSubstitution(LeftRightLogicalSubstitution):
+    """Substitution that returns 'or' of the input boolean values."""
 
+    def __init__(self, left: SomeSubstitutionsType, right: SomeSubstitutionsType) -> None:
+        """Create an OrSubstitution substitution."""
+        super().__init__(or_, left, right)
 
-@expose_substitution('any')
-class AnySubstitution(Substitution):
-    """
-    Substitutes to the string 'true' if at least one of the input arguments evaluates to true.
 
-    If none of the arguments evaluate to true, then this substitution returns the string 'false'.
-    """
+class ContainerSubstitution(Substitution):
+    """Substitution running a logical evaluation on the boolean values of the container."""
 
-    def __init__(self, *args: SomeSubstitutionsType) -> None:
+    def __init__(
+        self, func: Callable[[List[StrSomeValueType]], bool], *args: SomeSubstitutionsType
+    ) -> None:
         """
-        Create an AnySubstitution substitution.
+        Create a ContainerSubstitution substitution.
 
         The following string arguments evaluate to true: '1', 'true', 'True', 'on'
+        The following string arguments evaluate to false: '0', 'false', 'False', 'off'
```

Comment:
I would keep these (i.e., copy) in concrete classes, since users probably won't look at the base class' docstring.